rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -49,28 +71,113 @@
     private FlowNodeUtil() {
     }
 
-    public static long getNodeExecDuration(FlowNode node) {
-        List<FlowNode> childNodes = getChildNodes(node);
-        if (!childNodes.isEmpty()) {
-            long startTime = TimingAction.getStartTime(node);
-            long endTime = TimingAction.getStartTime(childNodes.get(childNodes.size() - 1));
+    public abstract static class CacheExtensionPoint implements ExtensionPoint {
+        public abstract Cache<String,List<FlowNode>> getExecutionCache();
+        public abstract Cache<String, RunExt> getRunCache();
+        public abstract Cache<FlowNode,String> getExecNodeNameCache();
+    }
 
-            if (startTime == 0) {
-                // The node is running and the time has not been marked on it yet.  Return 0 as the duration for now.
-                return 0;
+    // Used in testing where Jenkins is not running yet
+    private static final List<CacheExtension> FALLBACK_CACHES = Arrays.asList(new CacheExtension());
+
+    @Extension
+    public static class CacheExtension extends CacheExtensionPoint {
+
+        protected final Cache<String,List<FlowNode>> executionCache = CacheBuilder.newBuilder().weakValues().maximumSize(100).build();
+
+        // Larger cache of run data, for completed runs, keyed by flowexecution url, useful for serving info
+        // Actually can be used to serve Stage data too
+        // Because the RunExt caps the total elements returned, and this is fully realized, this is the fastest way
+        protected final Cache<String, RunExt> runData = CacheBuilder.newBuilder().maximumSize(1000).build();
+
+        protected final Cache<FlowNode,String> execNodeNameCache = CacheBuilder.newBuilder().weakKeys().expireAfterAccess(1, TimeUnit.HOURS).build();
+
+        public Cache<String,List<FlowNode>> getExecutionCache() {
+            return this.executionCache;
+        }
+        public Cache<String, RunExt> getRunCache() {
+            return this.runData;
+        }
+
+        public Cache<FlowNode,String> getExecNodeNameCache() {
+            return  this.execNodeNameCache;
+        }
+
+        public static List<CacheExtension> all() {
+            Jenkins myJenkins = Jenkins.getInstance();
+            if ( myJenkins == null) {
+                return FALLBACK_CACHES;
+            } else {
+                return myJenkins.getExtensionList(CacheExtension.class);
+            }
+        }
+    }
+
+    @CheckForNull
+    public static RunExt getCachedRun(FlowExecution ex) {
+        try {
+            if (ex != null) {
+                RunExt cachedRun = CacheExtension.all().get(0).getRunCache().getIfPresent(ex.getUrl());
```

Comment:
@kzantow Yeah, it's a little bit awkward, no lie.  Rather than addressing this in this PR though, let me take care of it with https://github.com/jenkinsci/pipeline-stage-view-plugin/pull/6 (I've added a TODO to my task list there).  I'm stripping out most of the caches there (no longer needed when the new FlowAnalyzer is added, except for the Run cache).